extensions/Worker.Extensions.Http/src/DefaultFromBodyConversionFeature.cs

```diff
@@ -0,0 +1,110 @@
+
+using System;
+using System.Linq;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Serialization;
+using Microsoft.Azure.Functions.Worker.Extensions.Http.Converters;
+using Microsoft.Azure.Functions.Worker.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Http
+{
+    internal class DefaultFromBodyConversionFeature : IFromBodyConversionFeature
+    {
+        public static IFromBodyConversionFeature Instance { get; } = new DefaultFromBodyConversionFeature();
+
+        public ValueTask<object?> ConvertAsync(FunctionContext context, Type targetType)
+        {
+            var requestDataResult = context.GetHttpRequestDataAsync();
+
+            if (requestDataResult.IsCompletedSuccessfully)
+            {
+                return ConvertRequestAsync(requestDataResult.Result, context, targetType);
+            }
+
+            return ConvertAsync(requestDataResult, context, targetType);
+        }
+
+        private async ValueTask<object?> ConvertAsync(ValueTask<HttpRequestData?> requestDataResult, FunctionContext context, Type targetType)
+        {
+            var requestData = await requestDataResult;
+            return await ConvertRequestAsync(requestData, context, targetType);
+        }
+
+        private ValueTask<object?> ConvertRequestAsync(HttpRequestData? requestData, FunctionContext context, Type targetType)
+        {
+            if (requestData is null)
+            {
+               throw new InvalidOperationException($"The '{nameof(DefaultFromBodyConversionFeature)} expects an '{nameof(HttpRequestData)}' instance in the current context.");
+            }
+
+            return ConvertBodyAsync(requestData, context, targetType);
+        }
+
+        private static ValueTask<object?> ConvertBodyAsync(HttpRequestData requestData, FunctionContext context, Type targetType)
+        {
+            object? result;
+
+            if (targetType == typeof(string))
+            {
+                result = requestData.ReadAsString();
+            }
+            else if (targetType == typeof(byte[]))
+            {
+                result = ReadBytes(requestData, context.CancellationToken);
+            }
+            else if (targetType == typeof(Memory<byte>))
+            {
+                Memory<byte> bytes = ReadBytes(requestData, context.CancellationToken);
+                result = bytes;
+            }
+            else if (HasJsonContentType(requestData))
+            {
+                ObjectSerializer serializer = requestData.FunctionContext.InstanceServices.GetService<IOptions<WorkerOptions>>()?.Value?.Serializer
+                 ?? throw new InvalidOperationException("A serializer is not configured for the worker.");
+
+                result = serializer.Deserialize(requestData.Body, targetType, context.CancellationToken);
```

Comment:
There is also a `DeserializeAsync` [method](https://learn.microsoft.com/en-us/dotnet/api/azure.core.serialization.objectserializer.deserializeasync?view=azure-dotnet) which returns a `ValueTask`.

Reply:
This is intentional. All code paths here are synchronous as  they're optimized for the mainline scenario using the default `HttpRequestData` implementation, where the body is loaded `MemoryStream`. 

Benchmarked this and the non-async method performed better for core scenarios.